base/math.jl

```diff
@@ -537,38 +537,61 @@ Stacktrace:
 [...]
 ```
 """
-hypot(x::Number, y::Number) = hypot(promote(x, y)...)
-function hypot(x::T, y::T) where T<:Number
-    ax = abs(x)
-    ay = abs(y)
-    if ax < ay
-        ax, ay = ay, ax
-    end
+function hypot(x::Number, y::Number)
+    # Specialized two argument version is faster than generic varargs version
+    ay, ax = minmax(hypot(x), hypot(y))
     if iszero(ax)
         r = ay / oneunit(ax)
     else
         r = ay / ax
     end
-
-    rr = ax * sqrt(1 + r * r)
-
-    # Use type of rr to make sure that return type is the same for
-    # all branches
+    r2 = r^2
+    rr = ax * sqrt(oneunit(r2) + r2)
     if isnan(r)
-        isinf(ax) && return oftype(rr, Inf)
-        isinf(ay) && return oftype(rr, Inf)
-        return oftype(rr, r)
+        if isinf(x) || isinf(y)
+            return oftype(rr, Inf)
+        else
+            return oftype(rr, r)
+        end
     else
         return rr
-    end
+   	end
 end
 
 """
     hypot(x...)
 
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+function hypot(x::Number...)
+    # speculatively try fast naive approach
+    s = _hypot_squared(x)
+    if isnan(s)
+        return any(isinf, x) ? oftype(sqrt(s), Inf) : sqrt(s) # IEEE 754
+    elseif isinf(s) || s ≤ oneunit(s) * _floatmin(one(s)) * (length(x) - 1)
+        # if overflow/underflow, try normalization
+        return _hypot_safe(x)
+    else
+        return sqrt(s)
+    end
+end
+
+hypot() = error("invalid hypot() call without arguments")
+hypot(x::Number) = abs(float(x))
+
+_hypot_squared(x::Tuple{Vararg{Number}}) = sum(abs2 ∘ float, x)
```

Comment:
Where is this function used?

Reply:
Oh, I see it above

Reply:
since it is only used once, why not just inline it?